compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1619,7 +1619,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
             tree.selector.removeAttachment(desugar.CheckIrrefutable) match {
               case Some(checkMode) if !sel.tpe.hasAnnotation(defn.UncheckedAnnot) =>
                 val isPatDef = checkMode == desugar.MatchCheck.IrrefutablePatDef
-                if !checkIrrefutable(sel, pat, isPatDef) && sourceVersion == `future-migration` then
+                if !checkIrrefutable(sel, pat, isPatDef) && sourceVersion.isAtLeast(`3.2`) && sourceVersion.isMigrating then
```

Comment:
I think this patch should only run upto that version where we `errorFrom` in the `gradualErrorOrMigrationWarning`, so here I introduce a new SourceVersion member `isBetween`:
```suggestion
                if !checkIrrefutable(sel, pat, isPatDef) && sourceVersion.isMigrating && sourceVersion.isBetween(`3.2`, `future`) then
```

```scala
def isBetween(low: SourceVersion, high: SourceVersion): Boolean =
  require(low.ordinal < high.ordinal)
  isAtLeast(low) && stable.ordinal <= high.ordinal
```

Reply:
perhaps also we should introduce a centralised place to control which versions each feature is enabled / warnFrom / errorFrom, so we don't need to synchronise version numbers everywhere when we enable something from `future`

Reply:
I also wonder if the "This patch can be rewritten automatically" message should mention the full range of source versions it applies to.